# Project 4 Problem 1

# Define lexnode

In [1]:
from collections import Counter
class LexNode:
    def __init__(self, val):
        self.val = val
        self.children = []
        # set the property so that we can differentiate the start node, normal(between) node and end of word node
        # 0: normal node
        # 1: start node
        # 2: end-of-word node
        self.property = 0
        self.idx=0
        
    # pretty print tree in command line
    def pretty_str(self, level=0):
        #print(self.val, level)
        if self.property==0:
            ret = "--"+self.val
        elif self.property==2:
            ret="--"+self.val+"\n"
        else:
            ret=self.val
        num_children=len(self.children)
        for i in range(num_children):
            if level!=0:
                if num_children>=2 and i!=0:
                    ret += "|  "*level+"|"+self.children[i].pretty_str(level+1)
                elif num_children==1 and self.children[0].val=="*":
                    return ret
                else:
                    ret += self.children[i].pretty_str(level+1) 
            elif level==0:
                if num_children>=2 and i!=0:
                    ret += "|\n"+"|"+self.children[i].pretty_str(level+1)
                else:
                    ret += self.children[i].pretty_str(level+1)
            
        return ret

    def get_max_level(self, level=0):
        if len(self.children) == 0:
            return level
        child_levels = [child.get_max_level(level=level + 1) for child in self.children]
        return max(child_levels)

# Build lextree

In [2]:
class BuildLextree:
    def __init__(self, dic_file_path):
        self.txt2words(dic_file_path)
        self.tree=LexNode('*')
        # dummy symbol for the root of the tree
        self.tree.property = 1
        n_words = len(self.words)
        word_lens = [len(w) for w in self.words]
        #find the longest word in this dictionary
        self.max_word_len = max(word_lens)
        # pad the words to the same length using '$', in order to avoid index out of range
        for n in range(n_words):
            self.words[n] = self.words[n].ljust(self.max_word_len,'$')
        print("There are {} words in this dictionary".format(len(self.words)))
        
    def txt2words(self,file_path):
        self.words=[]
        dictionary= open (file_path, "r")
        words=dictionary.readlines()
        for word in words:
            self.words.append(word.strip('\n'))
        
    def append_lex_node(self,parent, child):
        #This function just append the child node to the paretn node
        #It would check whether the parent is a LexNode!
        assert type(parent) is LexNode and type(child) is LexNode
        parent.children.append(child)
    
    def build_lextree(self):
        #this is the function to build the lextree from the self.words and root node "*"
        self._build_lextree(self.tree, self.words,0)
        
        
    def _build_lextree(self,node, words, i):
        '''
        This function is to build the tree based on the specific node. It is the basic part of this class
        input: 
        1.the node that need to add children
        2.the words that need to be reviewed
        3. the index that the character on which need to be reviewed
        4. the legthn of the longthest word
        '''
        # once, it reach the longest word and finished that part, we should stop
        if i == self.max_word_len:
            print("The Lextree has been build")
        #record the words that have reached their end
        to_be_removed=[]
        for word in words:
            # if current character is the end of a word, add it as a separate node even if the same node can be shared. This
            # is to ensure that every leaf represent exactly one word.
            if (i + 1 < self.max_word_len and word[i + 1] =='$') or i == self.max_word_len - 1:
                child = LexNode(word[i])
                child.property = 2
                #add "*" at the end of the word
                child.children.append(self.tree)
                self.append_lex_node(node, child)
                to_be_removed.append(word)
        #Now, remove the word that have reached the end
        if len(to_be_removed)!=0:
            for word in to_be_removed:
                words.remove(word)
        #for the rest words, let's append their left characters        
        n_words = len(words)
        # get the characters in j-th position
        chars = [words[n][i] for n in range(n_words)]
        counts = Counter(chars)
        #print(counts)
        for character, count in counts.items():
            # we only focus on this single character here
            child = LexNode(character)
            self.append_lex_node(node, child)
            #get the words that the i th character is the current character we look at
            words_to_be_review_for_this_character = [word for idx, word in enumerate(words) if word[i] == character]
            self._build_lextree(child, words_to_be_review_for_this_character, i + 1)

# Read the dictionary

In [3]:
dic_file_path="dict_1.txt"

# Build lexical tree based on the dictionary

In [4]:
buildlextree=BuildLextree(dic_file_path)
buildlextree.build_lextree()
lextree=buildlextree.tree

There are 6250 words in this dictionary


# Length of the longest word

In [5]:
longest_word_lenth=buildlextree.max_word_len

# Print the built-up lexical tree

In [6]:
print(lextree.pretty_str(0))

*--a
|
|--e
|
|--i
|
|--o
|
|--a--m
|  |--n
|  |--s
|  |--t
|  |--b--a--n--d--o--n
|  |  |--b--a--s
|  |  |  |--r--e--v--i--a--t--i--o--n
|  |  |--d--o--m--i--n--a--l
|  |  |--e--l--a
|  |  |  |--r--n--e--t--h--y
|  |  |--i--d--e--s
|  |  |--l--e
|  |  |--o--l--i--s--h--i--n--g
|  |  |  |--r--t--i--o--n--i--s--t--s
|  |  |  |--u--t
|  |  |--r--a--h--a--m
|  |  |  |--i--d--g--e
|  |  |--s--e--n--c--e--s
|  |  |  |--o--l--v--e--d
|  |  |  |--t--i--n--e--n--t
|  |  |--u--n--d--a--n--t--l--y
|  |--c--c--e--d--e--s
|  |  |  |  |--n--t--u--a--t--i--n--g
|  |  |  |  |--p--t
|  |  |  |--i--d--e--n--t
|  |  |  |--o--m--m--o--d--a--t--e--d
|  |  |  |  |  |--p--a--n--y
|  |  |  |  |  |  |--l--i--s--h--m--e--n--t
|  |  |  |  |--r--d
|  |  |  |  |--u--n--t--a--n--c--y
|  |  |  |--r--u--e
|  |  |  |--u--r--i--d--e
|  |  |--e--r--o
|  |  |--h--e--s
|  |  |--k--e--r--'--s
|  |  |  |--r--o--y--d
|  |  |--q--u--a--i--n--t
|  |  |  |  |--i--t
|  |  |--r--o--n--y--m
|  |  |  |  |--s--s
|  |  |--t--i--v--a

# Spell Checker

## Without beam

In [ ]:
import numpy as np
import copy
class SpellChecker:
    def __init__(self,beam,longest_word_lenth=45,tolerate_num_of_error=3):
        self.beam = beam
        self.lextree=None
        self.dist_fun=lambda *args: int(args[0] != args[1])
        self.longest_word_lenth=longest_word_lenth
        self.tolerate_num_of_error=tolerate_num_of_error

    def fit(self,lextree):
        self.lextree=lextree
        assert type(self.lextree) is LexNode
        self.nodes = []
        self.get_nodes(self.lextree)
        # get self.transitions
        self.transitions = {}
        self.get_chilren={}
        to_children={}
        n_nodes = len(self.nodes)
        self.word_ends = []
        # record the end idx of each words, therefore, at the end of the vertibe , we can get the costs of each word
        for i in range(n_nodes):
            n = self.nodes[i]
            #print(str(i)+"    "+n.val)
            if n.property == 2:
                self.word_ends.append(i)
            self.get_chilren[i]=[]
            # add transition if there is any. to get the parent node of current node
            if len(n.children) > 0:
                for child in n.children:
                    self.get_chilren[i].append(self.nodes.index(child))
                    self.transitions[self.nodes.index(child)] = i
        self.transitions[0]=0 # "start point repeating"
        
        
    def spell_check(self, text):
        # TODO: evaluate input
        pass
    def mute_nodes(self,r):
        self.muted_nodes[r]=1
        #mute it children
        to_mute_children=self.get_chilren[r]
        #print(to_mute_children)
        if len(to_mute_children)!=0:
            for child_idx in to_mute_children:
                if child_idx!=0:
                    self.mute_nodes(child_idx)
                
    def get_nodes(self, lexnode):
        lexnode.idx=len(self.nodes)
        self.nodes.append(lexnode)
        if lexnode.property==2:
            return
        for child in lexnode.children:
            self.get_nodes(child)
    
    def text_vertibe(self,x,longest=True,get_string=True):
        #initialize the input x by adding "*" at the beginning of x. 
        x = '*' + x
        # initialize cost matrix
        n_cols = len(x)
        n_rows = len(self.nodes)
        costs = np.full((n_rows, n_cols), np.inf)
        costs[0,:]=range(n_cols)
        
        n_nodes = len(self.nodes)
        self.muted_nodes=np.zeros([n_nodes])

        # fill cost matrix
        for c in range(n_cols):
            for r in range(1,n_rows):
                if not self.muted_nodes[r]:
                    current_cost=self.dist_fun(self.nodes[r].val,x[c])
                    if c==0:#if it is the initialization.
                        if self.transitions[r]==0:
                            costs[r,c]=current_cost
                        else:
                            #if it is the middle nodes, then the cost is based on it previous node
                            costs[r,c]=costs[self.transitions[r],c]+current_cost
                    elif c==1:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        else:
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost
                    else:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        else:
                            #that means they are different
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c],#update from current state previous node
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost


                    if costs[r,c]>=self.tolerate_num_of_error+len(x):
                        self.mute_nodes(r)
            #print(costs[:25,c])
        
        
        #find the minimum cost at the end of each word
        costs_for_words=[]
        for end_idx in self.word_ends:
            costs_for_words.append(costs[end_idx,-1])
        #print(costs_for_words[0:40])
        
        min_cost=min(costs_for_words)
        #print(min_cost)
        #get the index of minimum cost
        min_idxs=np.where(costs_for_words==min_cost)[0]
        #print(min_idxs)
        if get_string:
            result=self.nodes_to_word(min_idxs,longest=longest)
            return result
        else:
            return min_idxs, min_cost
        
    #Methods used in Problem 2
    def nodes_to_word(self,nodes,longest=True):
        results=[]
        for idx in nodes:
            idx=self.word_ends[idx]
            #if the min cost is the real end node, then we can get a real word
            result=""
            while self.transitions[idx]!=0:
                result=self.nodes[idx].val+result
                #print(self.nodes[idx].val)
                idx=self.transitions[idx]
            result=self.nodes[idx].val+result
            results.append(result)
            #print(results)
        if longest:
            #return the longest word in the candidate
            res = max(results, key=len, default='')
            return res
        return results
    
    def nodes_to_sentence(self,nodes):
        sentence=""
        for idx in nodes:
            idx=self.word_ends[idx]
            #if the min cost is the real end node, then we can get a real word
            result=""
            while self.transitions[idx]!=0:
                result=self.nodes[idx].val+result
                #print(self.nodes[idx].val)
                idx=self.transitions[idx]
            result=self.nodes[idx].val+result
            sentence+=result+" "
        
        return  sentence
    
    def unsegment_text_vertibe(self,x,longest=True,n_best=3):
        self.final_nodes,self.final_costs=self._unsegment_text_vertibe(x)
        
        to_visual_nodes_links=[]
        
        small_number = heapq.nsmallest(n_best,self.final_costs) 
        small_index = []
        for t in small_number:
            index = self.final_costs.index(t)
            small_index.append(index)
            self.final_costs[index] = 0
            to_visual_nodes_links.append( self.final_nodes[index])
            
        visualized_sentence=[]
        for node_links in to_visual_nodes_links:
            visualized_result=self.nodes_to_sentence(node_links)
            print(visualized_result)
            visualized_sentence.append(visualized_result)
        return visualized_sentence
        
    def _unsegment_text_vertibe(self,x,threshold=10):
        returned_words=[]
        returned_costs=[]
        #initialize the input x by adding "*" at the beginning of x. 
        x = '*' + x
        
        # initialize cost matrix
        #n_cols is just the to check length
        if len(x)>self.tolerate_num_of_error+self.longest_word_lenth:
            n_cols=self.tolerate_num_of_error+self.longest_word_lenth
        else:
            n_cols=len(x)
    
        
        n_rows = len(self.nodes)
        costs = np.full((n_rows, n_cols), np.inf)
        costs[0,:]=range(n_cols)
        
        n_nodes = len(self.nodes)
        self.muted_nodes=np.zeros([n_nodes])
        
        # fill cost matrix
        for c in range(n_cols):
            print("current  c is "+str(c))
            for r in range(1,n_rows):
                if not self.muted_nodes[r]:
                    current_cost=self.dist_fun(self.nodes[r].val,x[c])
                    if c==0:#if it is the initialization.
                        if self.transitions[r]==0:
                            costs[r,c]=current_cost
                        else:
                            #if it is the middle nodes, then the cost is based on it previous node
                            costs[r,c]=costs[self.transitions[r],c]+current_cost
                    elif c==1:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        else:
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost
                    else:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        else:
                            #that means they are different
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c],#update from current state previous node
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost


                    if costs[r,c]>=self.tolerate_num_of_error+len(x):
                        self.mute_nodes(r)
            #check whether there is already a new word complete
            #find the minimum cost at the end of each word
            if c!=0:
#                 print("current cost for all nodes is")
#                 print(costs[:,c])
#                 print(self.word_ends)
                costs_for_words=[]
                for end_idx in self.word_ends:
                    costs_for_words.append(costs[end_idx,c])
#                 print(costs_for_words)
                #print(costs_for_words[0:40])
                min_cost=min(costs_for_words)
#                 print("minimum cost at this point is")
#                 print(min_cost)
                #get the index of minimum cost
                min_idxs=np.where(costs_for_words==min_cost)[0]
                #when i reach the end of the string.
                if len(x)<=n_cols:
                    print("there are less than {} charactor left!".format(n_cols))
                    print("minimum cost at this point is {}".format(min_cost))
                    rest_words=[]
                    if c<n_cols-1:
                        #check if we could push the rest letter to be new word
                        rest_string=x[c+1:]
                        rest_words,rest_min_cost=self.text_vertibe(rest_string,get_string=False)      
                          
                    for i in range(len(min_idxs)):
                        word_end_idx=min_idxs[i]
                        if len(rest_words)!=0:# if the rest characters could be a word, record the word and its cost
                            min_cost=min_cost+rest_min_cost
                            for rest_words_node in rest_words:
                                returned_words.append([word_end_idx,rest_words_node])
                                returned_costs.append(min_cost)
                        else: #else just add what I have now
                            returned_words.append([word_end_idx])
                            returned_costs.append(min_cost)
            
                          
                # elif, it is still in the window part of the input string
                else:
                    next_words,next_costs=self._unsegment_text_vertibe(x[c+1:])
                    for i in range(len(min_idxs)):
                        word_end_idx=min_idxs[i]
                        new_nodes=[]
                        new_costs=[]
                        for next_node_id in range(len(next_costs)):
                            next_node=next_words[next_node_id]
                            next_cost=next_costs[next_node_id]
#                             print("next cost is {}".format(next_cost)  )
#                             print("next words is {}".format(next_node)  )
                            current_cost=min_cost+next_cost
#                             print("current_cost is {}".format(current_cost))
                            if current_cost<threshold:
                                new_costs.append(current_cost)
                                new_nodes.append([word_end_idx]+next_node)
#                             print("newnodes!")
#                             print(new_nodes)
                        returned_words+=new_nodes
                        returned_costs+=new_costs
            print("after get the new nodes, my returned words would be")
            print(returned_words)
            print(returned_costs)
        
        return   returned_words,returned_costs               

## With beam

In [21]:
import numpy as np
import copy
class SpellChecker:
    def __init__(self,beam,longest_word_lenth=45,tolerate_num_of_error=3):
        self.beam = beam
        self.lextree=None
        self.dist_fun=lambda *args: int(args[0] != args[1])
        self.longest_word_lenth=longest_word_lenth
        self.tolerate_num_of_error=tolerate_num_of_error

    def fit(self,lextree):
        self.lextree=lextree
        assert type(self.lextree) is LexNode
        self.nodes = []
        self.get_nodes(self.lextree)
        # get self.transitions
        self.transitions = {}
        self.get_children={}
        to_children={}
        n_nodes = len(self.nodes)
        self.word_ends = []
        # record the end idx of each words, therefore, at the end of the vertibe , we can get the costs of each word
        for i in range(n_nodes):
            n = self.nodes[i]
            #print(str(i)+"    "+n.val)
            if n.property == 2:
                self.word_ends.append(i)
            self.get_children[i]=[]
            # add transition if there is any. to get the parent node of current node
            if len(n.children) > 0:
                for child in n.children:
                    self.get_children[i].append(self.nodes.index(child))
                    self.transitions[self.nodes.index(child)] = i
        self.transitions[0]=0 # "start point repeating"
        
        
    def spell_check(self, text):
        # TODO: evaluate input
        pass
    def mute_nodes(self,r):
        self.muted_nodes[r]=1
        #mute it children
        to_mute_children=self.get_children[r]
        #print(to_mute_children)
        if len(to_mute_children)!=0:
            for child_idx in to_mute_children:
                if child_idx!=0:
                    self.mute_nodes(child_idx)
                
    def get_nodes(self, lexnode):
        lexnode.idx=len(self.nodes)
        self.nodes.append(lexnode)
        if lexnode.property==2:
            return
        for child in lexnode.children:
            self.get_nodes(child)
    
    def text_vertibe(self,x,longest=True,get_string=True):
        #initialize the input x by adding "*" at the beginning of x. 
        x = '*' + x
        # initialize cost matrix
        n_cols = len(x)
        n_rows = len(self.nodes)
        costs = np.full((n_rows, n_cols), np.inf)
        costs[0,:]=range(n_cols)
        
        n_nodes = len(self.nodes)
        self.muted_nodes=np.zeros([n_nodes])

        # fill cost matrix
        for c in range(n_cols):
            self.recursion(self.lextree, c, x, costs)
            #print(costs[:25,c])
        
        
        #find the minimum cost at the end of each word
        costs_for_words=[]
        for end_idx in self.word_ends:
            costs_for_words.append(costs[end_idx,-1])
        #print(costs_for_words[0:40])
        
        min_cost=min(costs_for_words)
        #print(min_cost)
        #get the index of minimum cost
        min_idxs=np.where(costs_for_words==min_cost)[0]
        #print(min_idxs)
        if get_string:
            result=self.nodes_to_word(min_idxs,longest=longest)
            return result
        else:
            return min_idxs, min_cost
        
    #Beam
    def recursion(self, lexnode,c,x,costs):
        r=lexnode.idx
        #r=self.nodes.index(lexnode)
        if not self.muted_nodes[r]:
            current_cost=self.dist_fun(lexnode.val,x[c])
            if c==0:#if it is the initialization.
                if self.transitions[r]==0:
                    costs[r,c]=current_cost
                else:
                    #if it is the middle nodes, then the cost is based on it previous node
                    costs[r,c]=costs[self.transitions[r],c]+current_cost
            elif c==1:
                if current_cost==0:
                    to_check=[costs[r,c-1]+1, # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                             ] 
                else:
                    to_check=[costs[r,c-1], # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                             ] 
                nnn=np.argmin(to_check)
                costs[r,c]=to_check[nnn]+current_cost
            else:
                if current_cost==0:
                    to_check=[costs[r,c-1]+1, # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                              costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                             ]
                else:
                    #that means they are different
                    to_check=[costs[r,c-1], # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                              costs[self.transitions[r],c],#update from current state previous node
                              costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                             ]
                nnn=np.argmin(to_check)
                costs[r,c]=to_check[nnn]+current_cost
            
            if costs[r,c]>=self.tolerate_num_of_error+len(x):
                    self.mute_nodes(r)

        #print(lexnode.val)
        if lexnode.property==2:
            return
        for child in lexnode.children:
            if costs[r,c]<=self.beam+1:
                self.recursion(child,c,x,costs)

# Initialize SpellChecker

In [22]:
spellchecker=SpellChecker(3,longest_word_lenth=9)
#Fit lextree into spellchecker
spellchecker.fit(lextree)

# Test SpellChecker

In [31]:
import datetime
print(datetime.datetime.now())
x="aple"
print(spellchecker.text_vertibe(x,True))
print(datetime.datetime.now())

2022-02-16 21:25:51.962510
able
2022-02-16 21:25:52.709426


# All possible results besides the longest word

In [10]:
x="axn"
spellchecker.text_vertibe(x,False)

['an']

# Read typo.txt and correct the wrong words

In [24]:
print(datetime.datetime.now())
file=open("typos.txt","r").readlines()
sentences=[]
for sentence in file:
    words=sentence.strip('\n').split(" ")
    if len(words)==1:
        #which means it is an empty line
        continue
    corrected_sentence=""
    for word in words:
        corrected_word=spellchecker.text_vertibe(word)
        #print(corrected_word)
        corrected_sentence+=corrected_word+" "
    #print(corrected_sentence)
    sentences.append(corrected_sentence)
print(datetime.datetime.now())

2022-02-16 17:17:33.596251
2022-02-16 17:21:35.790487


# Save the corrected sentence

In [36]:
#Corrected typo without beam
typos_without_beam=["once upon a time wigle brahmadatta as ing of benares the bodhisatta kamen to life at the foot of he himalayas as a donkey he grew strong and sturdy big of fraim well to do and lived by a degrave of the never annese in a forest haunt now at that them there was a crocodile darlings in the engles the crocodile's mate saw the great frame of the monkey and she conceived a longing to eke as harter so she seda to her lord serb i desire to feet the heart of that grace king of the monkeys ", 'dodd life sadye the crocodile i leed in the nature and heed liese on dry land how can we mach him ', "dyk hua or by crock see replays he met be kort if i don't get heed i shall die ", "all rate answered the crudele consoling ohr don't frable herself i have a plan i wild give coo his cart to feet ", 'so when the bodhisatta was sitting on the bank of the engles after taking a drink of water the crocodile dromon near and seide sir monkey what do your live on bad roots in this nolde family flags on the odier side of the ganges there is no end to the mangoes trees and labuja trees wit fruit sweet as coney is it not bear to kroc over lande have allen kinds of wildes fruit to mate ', 'lore crocodile the dundee inset the ganges is deep and waddle how shall i at across ', 'life coo want to gob i fill let ju sit upon my back and kady you over ', 'the monkey trusted am and agreed come here then seide the crocodile up on myer back with coo and up the monkey climbed but when the crocodile had swum a little wave he plunged the monkey under the later ', 'good friend you are letting me sink cried the minkel what is that or ', 'the brooded said you think i am carrying your out of pure good nature not a bit of it my wife has a clanging for your heart and i wanta to eve it to ohr to mate ', 'freier said the monkey it is nice of coo to tell me what if our cart were inside us when we go jumping among the tried tops it wild be all knocked to pieces ', "all where do you keep it asked the crocodile's ", 'the bodhisatta pointed out a big tried with clusters of crye crout standing not far off side said he there are our halts hanging on yonder fine tried ', "if you will show me your heart said the crocodile then i won't kill gob ", 'take meet to the trees then and i all point it out to your ', 'the crocodile brought him to the place the monkey leapt off his back and climbing hupe the nigg tree sat upon it oh silly crocodile saints he you ought that their were creatures that kept their garst in a treetop you are a hoole and i have outwitted you you may kept your crout to yourself lore body is great but you have no sesno ', 'and then to explain this idea he uttered the following stanzas ', 'roseapple jackfruit enqueso tops across the water their i see ', 'engulf of them i ant them not my nigg is good enough for me ', 'greet is yukon body berlin buts how much smaller is you with ', 'now go your ways serb crocodile for i eve had the best how with ', 'the crocodile feeling as said and miserable as if he had lost a thousand pieces of money ant back sorrowing to the please where he lived ']
corrected_withoutbeam=''
for x in typos_without_beam:
    corrected_withoutbeam+=x
    
corrected_withoutbeam

"once upon a time wigle brahmadatta as ing of benares the bodhisatta kamen to life at the foot of he himalayas as a donkey he grew strong and sturdy big of fraim well to do and lived by a degrave of the never annese in a forest haunt now at that them there was a crocodile darlings in the engles the crocodile's mate saw the great frame of the monkey and she conceived a longing to eke as harter so she seda to her lord serb i desire to feet the heart of that grace king of the monkeys dodd life sadye the crocodile i leed in the nature and heed liese on dry land how can we mach him dyk hua or by crock see replays he met be kort if i don't get heed i shall die all rate answered the crudele consoling ohr don't frable herself i have a plan i wild give coo his cart to feet so when the bodhisatta was sitting on the bank of the engles after taking a drink of water the crocodile dromon near and seide sir monkey what do your live on bad roots in this nolde family flags on the odier side of the gang

In [35]:
#Corrected typo with beam
typos_with_beam=['once upon a time wigle uncharacteristically as ing of uncharacteristically the uncharacteristically kamen to life at the foot of he uncharacteristically as a uncharacteristically he grew uncharacteristically and uncharacteristically big of uncharacteristically well to do and lived by a uncharacteristically of the uncharacteristically uncharacteristically in a uncharacteristically uncharacteristically now at that them uncharacteristically was a uncharacteristically uncharacteristically in the uncharacteristically the uncharacteristically uncharacteristically saw the uncharacteristically uncharacteristically of the uncharacteristically and she uncharacteristically a uncharacteristically to eke as uncharacteristically so she seda to her lord serb i uncharacteristically to feet the uncharacteristically of that uncharacteristically king of the uncharacteristically ', 'dodd life sadye the uncharacteristically i leed in the uncharacteristically and heed uncharacteristically on dry land how can we mach him ', 'dyk hua or by crock see uncharacteristically he met be kort if i uncharacteristically get heed i uncharacteristically die ', 'all rate uncharacteristically the uncharacteristically uncharacteristically ohr uncharacteristically uncharacteristically uncharacteristically i have a plan i wild give coo his cart to feet ', 'so when the uncharacteristically was uncharacteristically on the bank of the uncharacteristically after uncharacteristically a drink of water the uncharacteristically dromon near and seide sir uncharacteristically what do your live on bad uncharacteristically in this nolde uncharacteristically uncharacteristically on the uncharacteristically side of the uncharacteristically uncharacteristically is no end to the uncharacteristically uncharacteristically and uncharacteristically uncharacteristically wit uncharacteristically uncharacteristically as coney is it not bear to kroc uncharacteristically lande have allen uncharacteristically of uncharacteristically uncharacteristically to mate ', 'lore uncharacteristically the uncharacteristically uncharacteristically the uncharacteristically is uncharacteristically and uncharacteristically how shall i at uncharacteristically ', 'life coo want to gob i fill let ju sit upon my back and kady you over ', 'the uncharacteristically uncharacteristically am and uncharacteristically come here then seide the uncharacteristically up on myer back with coo and up the uncharacteristically uncharacteristically but when the uncharacteristically had swum a uncharacteristically wave he uncharacteristically the uncharacteristically under the uncharacteristically ', 'good uncharacteristically you are uncharacteristically me uncharacteristically uncharacteristically the uncharacteristically what is that or ', 'the uncharacteristically said you uncharacteristically i am uncharacteristically your out of uncharacteristically uncharacteristically uncharacteristically not a bit of it my wife has a uncharacteristically for uncharacteristically uncharacteristically and i uncharacteristically to eve it to ohr to mate ', 'uncharacteristically said the uncharacteristically it is nice of coo to tell me what if our cart uncharacteristically uncharacteristically us when we go uncharacteristically uncharacteristically the tried tops it wild be all uncharacteristically to uncharacteristically ', 'all where do you keep it asked the uncharacteristically ', 'the uncharacteristically uncharacteristically out a big tried with uncharacteristically of crye uncharacteristically uncharacteristically not far off side uncharacteristically he uncharacteristically are our uncharacteristically uncharacteristically on uncharacteristically fine tried ', 'if you uncharacteristically uncharacteristically me your uncharacteristically said the uncharacteristically then i uncharacteristically kill gob ', 'uncharacteristically meet to the uncharacteristically then and i all uncharacteristically it out to your ', 'the uncharacteristically uncharacteristically him to the uncharacteristically the uncharacteristically uncharacteristically off his back and uncharacteristically hupe the nigg tree sat uncharacteristically it oh uncharacteristically uncharacteristically uncharacteristically he you uncharacteristically that uncharacteristically were uncharacteristically that kept uncharacteristically uncharacteristically in a uncharacteristically you are a uncharacteristically and i have uncharacteristically you you may kept your uncharacteristically to uncharacteristically lore body is uncharacteristically but you have no uncharacteristically ', 'and uncharacteristically to uncharacteristically this uncharacteristically he uncharacteristically the uncharacteristically uncharacteristically ', 'uncharacteristically uncharacteristically uncharacteristically tops uncharacteristically the water uncharacteristically i see ', 'uncharacteristically of them i ant them not my nigg is uncharacteristically uncharacteristically for me ', 'uncharacteristically is yukon uncharacteristically uncharacteristically buts how uncharacteristically uncharacteristically is you with ', 'now go uncharacteristically ways serb uncharacteristically for i eve had the uncharacteristically how with ', 'the uncharacteristically uncharacteristically as said and uncharacteristically as if he had lost a uncharacteristically uncharacteristically of uncharacteristically ant uncharacteristically uncharacteristically to the uncharacteristically where he lived ']
corrected_withbeam=''
for x in typos_with_beam:
    corrected_withbeam+=x
    
corrected_withbeam

'once upon a time wigle uncharacteristically as ing of uncharacteristically the uncharacteristically kamen to life at the foot of he uncharacteristically as a uncharacteristically he grew uncharacteristically and uncharacteristically big of uncharacteristically well to do and lived by a uncharacteristically of the uncharacteristically uncharacteristically in a uncharacteristically uncharacteristically now at that them uncharacteristically was a uncharacteristically uncharacteristically in the uncharacteristically the uncharacteristically uncharacteristically saw the uncharacteristically uncharacteristically of the uncharacteristically and she uncharacteristically a uncharacteristically to eke as uncharacteristically so she seda to her lord serb i uncharacteristically to feet the uncharacteristically of that uncharacteristically king of the uncharacteristically dodd life sadye the uncharacteristically i leed in the uncharacteristically and heed uncharacteristically on dry land how can w

In [42]:
#Correct stoty
f=open("storycorrect.txt")
lines = f.readlines()
temp=""
for line in lines:
    temp+=line
f.close()

temp

'Once upon a time, while Brahmadatta was king of Benares, the Bodhisatta came to life at the foot of the Himalayas as a monkey. He grew strong and sturdy, big of frame, well to do, and lived by a curve of the river Ganges in a forest haunt. Now at that time there was a crocodile dwelling in the Ganges. The crocodile\'s mate saw the great frame of the monkey, and she conceived a longing to eat his heart. So she said to her lord, "Sir, I desire to eat the heart of that great king of the monkeys!"\n\n"Good wife," said the crocodile, "I live in the water and he lives on dry land. How can we catch him?"\n\n"By hook or by crook," she replied, "he must be caught. If I don\'t get him, I shall die."\n\n"All right," answered the crocodile, consoling her, "don\'t trouble yourself. I have a plan. I will give you his heart to eat."\n\nSo when the Bodhisatta was sitting on the bank of the Ganges, after taking a drink of water, the crocodile drew near, and said, "Sir Monkey, why do you live on bad fr

# Accuracy

In [43]:
def distance(x, y):
    return int(x!=y)

#Calculate the accuracy of the corrected result
def acc_checker1(inp, tem, distance):
    M, N=len(tem), len(inp)
    #print(M, N)
    cost=np.zeros((M, N))
    
    cost[0, 0] = distance(inp[0], tem[0])
    for i in range(1, M):
        cost[i, 0] = cost[i-1, 0] + distance(inp[0], tem[i])

    for j in range(1, N):
        cost[0, j] = cost[0, j-1] + distance(inp[j], tem[0])

    # Populate rest of cost matrix within window
    for i in range(1, N):
        for j in range(1, M):
        
            choices = cost[j, i-1], cost[j-1, i-1], cost[j-1, i]
            #print(i, j)
            cost[j, i] = min(choices) + distance(inp[i], tem[j])

    # Return DTW distance given window 
    return cost[-1, -1]

In [49]:
#Cost between the corrected typos and the correct story
cost_nobeam=acc_checker1(corrected_withoutbeam, temp, distance)
cost_withbeam=acc_checker1(corrected_withbeam, temp, distance)
cost_nobeam, cost_withbeam

(564.0, 3237.0)

In [51]:
#Accuracy of the corrected result
acc_nobeam=1-cost_nobeam/len(temp)
acc_withbeam=1-cost_withbeam/len(corrected_withbeam)
print('Accuracy without beam: ', acc_nobeam, 'Accuracy with beam 3: ', acc_withbeam)

Accuracy without beam:  0.8220258756705585 Accuracy with beam 3:  0.40364775239498896


#### The accuracy without beam width is about 82% while the accuracy with beam width 3 is only about 40%.